File: 03-can-sum/01-t2.py


def canSum(lst, target):
    if target == 0:
        return True

    for num in lst:
        rem = target - num

        if rem < 0:
            continue

        if canSum(lst, rem):
            return True

    return False


if __name__ == '__main__':
    print(canSum([2, 3], 7))
    print(canSum([5, 3, 4, 7], 7))
    print(canSum([2, 4], 7))
    print(canSum([2, 3, 5], 8))
    print(canSum([7, 14], 300))

    # True, True, False, True, False


# In this solution, we are bruteforcing more than necessary.
# Since we are iterating over and over, it takes more time.

File: 03-can-sum/01-t2.cpp
#include <iostream>
#include <vector>

using namespace std;

bool canSumHelper(vector<int> &lst, int target)
{
    if (target < 0)
        return false;

    if (target == 0)
        return true;

    for (auto num : lst)
        if (canSumHelper(lst, target - num))
            return true;

    return false;
}

bool canSum(vector<int> lst, int target)
{
    return canSumHelper(lst, target);
}

int main()
{
    cout << canSum(vector<int>{2, 3}, 7) << endl;
    cout << canSum(vector<int>{5, 3, 4, 7}, 7) << endl;
    cout << canSum(vector<int>{2, 4}, 7) << endl;
    cout << canSum(vector<int>{2, 3, 5}, 8) << endl;
    cout << canSum(vector<int>{7, 14}, 300) << endl;

    // true, true, false, true, false

    return 0;
}